src/zkevm_specs/evm_circuit/execution/error_write_protection.py

```diff
@@ -0,0 +1,73 @@
+from zkevm_specs.evm_circuit.table import CallContextFieldTag
+
+from ...util import FQ
+from ..instruction import Instruction
+from ..opcode import Opcode
+
+
+# There are some op codes which modify state.
+# There are `[SSTORE, CREATE, CREATE2, CALL, SELFDESTRUCT, LOG0, LOG1, LOG2, LOG3, LOG4]`.
+# When execution call context is read only (static call) and internal call,
+# these op codes running will encounter write protection error
+def error_write_protection(instruction: Instruction):
+    # retrieve op code associated to oog constant error
+    opcode = instruction.opcode_lookup(True)
+    (
+        is_sstore,
+        is_create,
+        is_create2,
+        is_call,
+        is_selfdestruct,
+        is_log0,
+        is_log1,
+        is_log2,
+        is_log3,
+        is_log4,
+    ) = instruction.multiple_select(
+        opcode,
+        (
+            Opcode.SSTORE,
+            Opcode.CREATE,
+            Opcode.CREATE2,
+            Opcode.CALL,
+            Opcode.SELFDESTRUCT,
+            Opcode.LOG0,
+            Opcode.LOG1,
+            Opcode.LOG2,
+            Opcode.LOG3,
+            Opcode.LOG4,
+        ),
+    )
+
+    # Spec 1. opcode must be [SSTORE, CREATE, CREATE2, CALL, SELFDESTRUCT, LOG0, LOG1, LOG2, LOG3, LOG4].
+    instruction.constrain_equal(
+        is_sstore
+        + is_create
+        + is_create2
+        + is_call
+        + is_selfdestruct
+        + is_log0
+        + is_log1
+        + is_log2
+        + is_log3
+        + is_log4,
+        FQ(1),
+    )
+
+    # Spec 2.
+    # current call must be an internal call
+    is_root = instruction.call_context_lookup(CallContextFieldTag.IsRoot)
+    instruction.constrain_equal(is_root, FQ(0))
```

Comment:
Is there a way to have a Root call with static=true?  If that's not possible, maybe this constraint is redundant?

Reply:
I'm not sure I understand it fully. 

It's not possible bcs if it's a root call and static=true, then this case must be handled by `STATICCALL` opcode. If the root call with any one of `[SSTORE, CREATE, CREATE2, CALL, SELFDESTRUCT, LOG0, LOG1, LOG2, LOG3, LOG4]` opcodes, then  `static` must be false and wiil be handled by `[SSTORE, CREATE...]` instead of `error_write_protection`  in the call context .
Aka. any call contexts with static=true at least having a `STATICCALL` ahead. Is that correct?

Reply:
fixed in [672f4f9](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/462/commits/672f4f9b9c3e9be3676620e2b6ac7f85e42efb70)

Reply:
> Aka. any call contexts with static=true at least having a `STATICCALL` ahead. Is that correct?

I think so!  And since static=true needs a parent `STATICCALL`, then it can't be root (because a root call has no parent callop opcode)